File: python/leetcode.403.py
class Solution(object):
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        n = len(stones)
        dp = [[None for j in xrange(n)] for i in xrange(n)]
        import bisect

        def dfs(i, k):
            if dp[i][k] != None:
                return dp[i][k]
                
            if i == n - 1:
                return True

            pos = stones[i]
            ans = False
            for d in [k - 1, k, k + 1]:
                if d > 0:
                    new_pos = pos + d
                    index = bisect.bisect_left(stones, new_pos)
                    if index < n and stones[index] == new_pos:
                        ans |= dfs(index, d)
                        # Cut
                        if ans:
                            break

            dp[i][k] = ans
            return ans

        if n == 1:
            return True
        else:
            if stones[1] - stones[0] == 1:
                return dfs(1, 1)
            else:
                return False

sln = Solution()
print sln.canCross([0,1])
print sln.canCross([0,1,3,5,6,8,12,17])
print sln.canCross([0,1,2,3,4,8,9,11])

File: cpp/leetcode.403.cpp

#include <cstdio>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <cstring>

using namespace std;

char dp[2020][2020];

class Solution {
public:
    bool dfs(vector<int>& stones, int i, int k){
        int n = stones.size();
        if(dp[i][k] != -1){
            return dp[i][k];
        }
        if(i == n - 1){
            return true;            
        }
        int pos = stones[i];
        bool ans = false;
        for(int d = k - 1; d <= k + 1; d++){
            if(d > 0){
                int new_pos = pos + d;
                auto iter = lower_bound(stones.begin(), stones.end(), new_pos);
                int index = iter - stones.begin();
                if(iter != stones.end() && stones[index] == new_pos){
                    ans = ans || dfs(stones, index, d);
                }
            }
        }
        dp[i][k] = ans;
        return ans;
    }
    bool canCross(vector<int>& stones) {
        int n = stones.size();
        memset(dp, -1, sizeof dp);
        if (n == 1){
            return true;
        }
        else{
            if (stones[1] - stones[0] == 1){
                return dfs(stones, 1, 1);
            } 
            else{
                return false;
            } 
        }
    }
};

int main(){
    Solution sln;
    vector<int> vec ;
    vec = {0,1,2,3,4,8,9,11};
    vec = {0,1,3,5,6,8,12,17};
    printf("%s\n", sln.canCross(vec) ? "true": "false");
}